Redesign viewstate package

sergefdrv opened the pull request:
This pull request significantly changes interface and implementation of `core/internal/viewstate` package. Taking into advantage some conclusions about protocol invariants and requirements, the component's interface, usage, and implementation are greatly simplified. For more details, see individual commit messages.

As a side-effect, this pulls request also resolves #110.

Commit:
core/internal/viewstate: Simplify description

The package's documentation is unnecessarily complicated. It can be
simplified by slightly changing terminology. Specifically, referring
to 'normal' and 'view-change' modes only adds excessive details and
distracts from the main purpose of package. Conversely, introducing
notions of 'current', 'expected', and 'requested' view numbers helps
to convey the package's purpose more clearly and concisely.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core/internal/viewstate: Use RWMutex to hold view

Although the current implementation seems to handle synchronization
correctly, it is overly complicated and might still have some bugs,
whereas standard RWMutex can do the same job. Take advantage of
RWMutex and simplify the code.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core/internal/viewstate: Add mocks for testing

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core/internal/viewstate: Use YAML to define test cases

Test cases are easier to follow when their definitions are compact and
come close to each other.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core: Remove code duplication in TestMakeViewMessageProcessor

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core: Pass view number as parameter to requestApplier

Current implementation of requestApplier performs all the steps that
will be required to restart consensus for a pending request when a
replica completes view change. However, to be used for that purpose,
it should not try acquiring and holding active view, since that must
be done by the caller. To prepare this piece of code for future usage
and simplify further changes, acquire the active view number in
requestProcessor, and pass it to requestApplier as parameter.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core: Use view state component directly

There is little benefit in defining aliases for methods of view state
component. Use the view state component directly instead.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

nhoriguchi commented on core/internal/viewstate/view-state.go:
```diff
@@ -17,276 +17,89 @@
-	FinishViewChange(newView uint64) (ok bool, release func())
+	HoldView() (current, expected uint64, release func())
+	AdvanceExpectedView(view uint64) (ok bool, release func())
+	AdvanceCurrentView(view uint64) (ok, active bool, release func())
```
The first return value means that expected view or current view is advanced or not, so renaming `ok` with `advanced` looks more self-descriptive to me.

sergefdrv replied:
The idea behind the name `ok` was to indicate if the requested operation - advancing the view number - was successfully performed by the invocation. Renaming it to `advanced` would indeed make it even more straightforward.

ynamiki replied:
I'm fine with both `ok` and `advanced`. Using `ok` for the meaning is natural for me and I have seen many usages.

sergefdrv replied:
Then I'd stick to `ok`. Thanks,

nhoriguchi commented on core/internal/viewstate/view-state.go:
```diff
@@ -17,276 +17,89 @@
-
-	// Last started view change
-	lastNewView uint64
+	sync.RWMutex
```
Introducing `RWMutex` here looks very effective to me, because accesses to current/expected view are mostly read accesses.

nhoriguchi commented on core/request.go:
```diff
@@ -166,19 +168,23 @@ func makeRequestProcessor(captureSeq requestSeqCapturer, pendingReq requestlist.
+		defer releaseView()
+
+		if currentView != expectedView {
+			return true, nil
```
Returning `true` for `new` means that MessageStreamHandler considers that the request is properly handled (leaving "Handled %s" log), but actually in this case the request seems not to be processed. So could that cause confusion?

sergefdrv replied:
I think this may be a grey area for that return value. On one hand, the request identifier is captured and the request is added to the pending request set before checking this. On the other hand, the replica has not started timers, nor generated a Prepare message at this point. However, the request identifier was captured as new, and the message was processed as much as the replica could do in its current state...

nhoriguchi replied:
We expect that we will restart to process the request in pending request set after view change completes in some mechanism (not implemented yet), so introducing another branch to return silently might be an option.

sergefdrv replied:
We could refine diagnostic message logging later, before making the first release. So far, we cannot return `new == false` here because the message is new and not dropped. On the other hand, returning `new == true` causes `Handled ...` message with _debug_ level.

nhoriguchi commented on core/message-handling.go:
```diff
@@ -417,29 +417,38 @@ func makeUIMessageProcessor(captureUI uiCapturer, processViewMessage viewMessage
+			if messageView < currentView {
 				return false, nil
+			} else if messageView > currentView {
+				return false, fmt.Errorf("Message refers to unexpected view")
```
It's a bit unclear to me why `messageView > currentView` case is considered as error and `messageView < currentView` case is not. Maybe that will get clearer when more view-change patches are submitted later, but it's helpful to leave some comment for background of the choice.

sergefdrv replied:
Good point. If the message refers to an old view then it is simply considered as delayed due to asynchrony in communication; so it gets ignored. However, if the message refers to a future view then the sender is considered faulty, and the error is reported. The latter condition is based on two arguments. 

First, any correct replica, when it starts participating in a new view, must make sure that all other correct replicas will do the same, as described in commit message of "core: Do not wait for active view". 

Second, we use reliable FIFO channels for communication and handle messages received from the same channel sequentially. This implies that messages are processed in the same order as they are sent by the peer. 

Thus, putting it all together, when a replica processes a normal-case message from a correct replica, it must have already transitioned into the corresponding view (or a later view), because the sender must have ensured that before sending the message.

sergefdrv replied:
I am thinking of adding a comment before returning the error, like: "Correct peer replicas ensure that other replicas transition into a new view before sending them any normal-case message for that view."

nhoriguchi replied:
> First, any correct replica, when it starts participating in a new view, must make sure that all other correct replicas will do the same, as described in commit message of "core: Do not wait for active view".
> 
> Second, we use reliable FIFO channels for communication and handle messages received from the same channel sequentially. This implies that messages are processed in the same order as they are sent by the peer.

I've overlooked that message order is guaranteed to be preserved by gRPC layer. I agree that the 2 points explain the reason well. Thank you.

> I am thinking of adding a comment before returning the error, like: "Correct peer replicas ensure that other replicas transition into a new view before sending them any normal-case message for that view."

Sounds nice.

sergefdrv replied:
> I've overlooked that message order is guaranteed to be preserved by gRPC layer.

This is also an [API requirement](https://godoc.org/github.com/hyperledger-labs/minbft/api#ReplicaConnector): "The connection guarantees authenticated destination, eventual delivery, preserving integrity and the original order of messages."

nhoriguchi reviewed (commented):
Thanks for moving forward to implement view-change. I left some comments/questions but mainly I do agree with whole ideas on this proposal.

ynamiki commented on core/internal/viewstate/view-state.go:
```diff
@@ -101,21 +72,14 @@ type State interface {
-	lastActiveView uint64
+	currentView   uint64
+	expectedView  uint64
+	requestedView uint64
```
Even we increment a view 1 million times per second, 585 thousand years are needed to consume `uint64` value. So we don't need to consider exhausting these values now (and also in the future).

sergefdrv replied:
Good observation :+1:

ynamiki reviewed (approved):
Thanks for the improvement. I left some comments, but they are not requests for changes.

sergefdrv commented:
Summary of [changes](https://github.com/hyperledger-labs/minbft/compare/dbf8db0460ba2ca17a2acf30a4347a2044f6e139..31ab4a9d9a8db1d39e817f1fb95f68917f4b78a3): 

- Added comment to explain "Message refers to unexpected view" error.

Commit:
core: Always process messages for current view

Obeying invariants of the protocol logic, it is safe to take into
account messages referring to the current view, even if the view is
not active, i.e. view change is in progress. This will allow to use
the same logic for processing both messages received during a stable
view and those received after view change was triggered in the
replica, including messages embedded in NewView and ViewChange
messages. However, a replica should skip some actions when processing
messages referring to an inactive current view, e.g. it should not
generate any message for the inactive view.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core: Do not wait for active view

When a correct replica completes view change and begins participating
in a new view, it must make sure other correct replicas will follow.
Otherwise, a faulty new primary might make some correct replicas
waiting to complete view change indefinitely long, e.g. by failing to
deliver its NewView message to them.

This conclusion allows to simplify code by replacing methods of view
state component that wait for active view with a method that just
obtains the actual values of current and expected view numbers.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core/internal/viewstate: Do not maintain requested view number

There is no need to maintain the last requested view number in this
component. This value can be maintained by the code responsible for
generating ReqViewChange messages to avoid generating duplicate
messages, whereas HoldView method can be used to synchronise
generation of ReqViewChange message with beginning of view change.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core/internal/viewstate: Rename updating methods

Now, the name of FinishViewChange method does not reflect its changed
semantics well, because the method returns "ok" whenever the current
view number is updated, but not only when view change is finished.
Rename both StartViewChange and FinishViewChange to better reflect
their semantics.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core/internal/viewstate: Do not update expected view in FinishViewChange

This change simplifies the semantics and implementation of the method,
whereas the protocol logic will benefit from starting view change as a
separate step before proceeding to complete it. The rationale is as
follows.

Whenever a replica decides to stop participating in an active view and
begin view change operation, it must make sure other correct replicas
will eventually do the same. Otherwise, other correct replicas might
still continue participating in the old view for arbitrary long time,
and the replica that started view change alone will stop making any
progress. Always starting view change as a separate step will help to
reuse the code that ensures the aforementioned liveness requirement.

Another benefit of always starting view change as a separate step is
that each subsequent view will be terminated with a ViewChange
message. This will make validation of message logs and processing of
new-view certificates easier and less error-prone.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>

Commit:
core/internal/viewstate: Indicate any update of current view number

A replica can safely process any valid NewView message referring to a
view number that is greater than the replica's current view number.
Doing so will allow for using the same code to process both directly
received NewView messages and those embedded into other messages.
However, a replica should only begin participating in a new view if
that view has become active.

Augment FinishViewChange method with additional return value to
indicate both if the current view number was updated and if the view
has become active.

Signed-off-by: Sergey Fedorov <sergey.fedorov@neclab.eu>
